Alerting: Migrate Silences to RTK Query

tomratcliffe opened the pull request:
**What is this feature?**

Migrates the API logic in Silences table/editor to use RTK Query instead, and tidies up appropriately afterwards.

Moves fetching logic into necessary components so we don't have to search through the API response to get an existing silence by ID. Uses `/:id` endpoint for fetching an individual silence, and uses tag invalidation/polling to fetch silences as needed

**Why do we need this feature?**

Simplifying the future work for Silences.

tomratcliffe committed:
Add silences RTK Query API

tomratcliffe committed:
Add mock server handlers necessary for Silences

tomratcliffe committed:
Restructure Silences components for better separation of concerns

tomratcliffe committed:
Remove unused actions/reducers

tomratcliffe committed:
Add handlers to mock server

tomratcliffe committed:
Refactor Silences test

tomratcliffe committed:
Remove unused `fetchAlerts` method

tomratcliffe committed:
Add correct tag invalidation after creating a silence

tomratcliffe committed:
Misc tidy up

tomratcliffe commented on public/app/features/alerting/unified/Silences.test.tsx:
```diff
@@ -285,48 +252,20 @@ describe('Silence edit', () => {
 
       await userEvent.click(ui.editor.submit.get());
 
-      await waitFor(() =>
```
I've removed some of the specific assertions here as we were checking that we called the API in a certain way. Instead I'm checking that the table is then rendered again afterwards - anyone have some other ideas of something to assert against here? I was trying to avoid continuing a pattern where we checked what MSW received 🤷

konrad147 replied:
We need to migrate these `api` calls assertions to check UI changes after API actions

tomratcliffe replied:
I added it back in by checking the server events but I'm not keen on using that approach much, so I marked the new helper method as deprecated immediately to make it clear in usages 😁 

Ideally we would have a stateful server for the tests (using MSW Data or something), and then we can assert its on the page correctly after without caring about the API's received body, or we could use cache updates in RTKQ and assert on the page state after that

tomratcliffe commented on public/app/features/alerting/unified/Silences.tsx:
I've moved the fetching logic into the corresponding components, so the list view and edit views will only fetch what they require (even though in practice, you have to go through the list view to edit a silence right now).

tomratcliffe commented on public/app/features/alerting/unified/components/silences/SilencesEditor.tsx:
```diff
@@ -97,24 +95,24 @@ const getDefaultFormValues = (searchParams: URLSearchParams, silence?: Silence):
 
-export const SilencesEditor = ({ silence, alertManagerSourceName }: Props) => {
+export const SilencesEditor = ({ silenceId, alertManagerSourceName }: Props) => {
+  const history = useHistory();
```
(Maybe todo/note to self: sort out useHistory vs locationService? What's going on here?)

tomratcliffe replied:
I've refactored to use that now - I had to make some changes to the test render method that I recently added (but I expected this might have to happen as we start adopting the method and using it for different cases)

tomratcliffe commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -29,12 +32,36 @@ export interface SilenceTableItem extends Silence {
+  );
+
+  const mimirLazyInitError =
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Any suggestions/patterns we're using elsewhere to handle errors? BackendSrv (I think) is giving it back to us always as `unknown`

tomratcliffe replied:
(now using `stringifyErrorLike` as per Gilles' suggestion elsewhere)

tomratcliffe commented on public/app/features/alerting/unified/mocks/server/handlers.ts:
```diff
@@ -4,7 +4,60 @@
 export const alertmanagerChoiceHandler = (response = defaultAlertmanagerChoiceResponse) =>
   http.get('/api/v1/ngalert', () => HttpResponse.json(response));
+
+const alertmanagerAlertsListHandler = () =>
+  http.get('/api/alertmanager/:datasourceUid/api/v2/alerts', () =>
+    HttpResponse.json([
+      mockAlertmanagerAlert({
+        labels: { foo: 'bar', buzz: 'bazz' },
+        status: { state: AlertState.Suppressed, silencedBy: ['12345'], inhibitedBy: [] },
+      }),
+      mockAlertmanagerAlert({
+        labels: { foo: 'bar', buzz: 'bazz' },
+        status: { state: AlertState.Suppressed, silencedBy: ['12345'], inhibitedBy: [] },
+      }),
+    ])
+  );
+
```
I didn't generalise this here, not sure what we want the mock API to return as a baseline - any ideas?

konrad147 replied:
In the best scenario this should be tied to datasources setup. 
404, at least in theory should be pretty rare (only Cortex responds this way)
For Prometheus and Mimir datasources we should provide valid JSON. GMA doesn't send this request

tomratcliffe commented on public/app/features/alerting/unified/mocks/server/handlers.ts:
```diff
@@ -4,7 +4,60 @@
+// Datasources //
+/////////////////
+
+// TODO: Add more accurate endpoint responses as tests require
+const datasourceBuildInfoHandler = () =>
+  http.get('/api/datasources/proxy/uid/:datasourceUid/api/v1/status/buildinfo', () => HttpResponse.json({}));
```
Likewise, not sure what we want to return here as when I followed the logic through on my local it was always 404'ing but not caring, so I don't know what we care about in particular here.

konrad147 replied:
In the best scenario this should be tied to datasources setup.
404, at least in theory should be pretty rare (only Cortex responds this way)
For Prometheus and Mimir datasources we should provide valid JSON. GMA doesn't send this request

tomratcliffe replied:
What would you suggest as a change in this PR? Or could we start generalising this as we move other tests to the consolidated MSW server?

konrad147 replied:
Let's postpone the decision. Not an issue in this PR

tomratcliffe committed:
Add comment explaining reset in silences editor

tomratcliffe commented on public/app/features/alerting/unified/Silences.tsx:
```diff
@@ -1,59 +1,18 @@
-      }
-    }
-    fetchAll();
-    const interval = setInterval(() => fetchAll, SILENCES_POLL_INTERVAL_MS);
-    return () => {
-      clearInterval(interval);
-    };
```
I don't think this was working before - I didn't see it refetch silences periodically

konrad147 replied:
You're right, it doesn't work. The function passed to `setInterval` returns `fetchAll` function instead of invoking it, so no requests are made

tomratcliffe replied:
Well now it works, and it does `refetchOnFocus` as well 🙌

konrad147 commented on public/app/features/alerting/unified/api/alertingApi.ts:
```diff
@@ -35,6 +35,8 @@ export const alertingApi = createApi({
     'AlertmanagerChoice',
     'AlertmanagerConfiguration',
+    'AlertmanagerAlerts',
+    'AlertSilences',
```
I'd rename to `AlertmanagerSilences`

tomratcliffe replied:
Done! Thanks

konrad147 commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -29,12 +32,36 @@ export interface SilenceTableItem extends Silence {
 
-const SilencesTable = ({ silences, alertManagerAlerts, alertManagerSourceName }: Props) => {
+const SilencesTable = ({ alertManagerSourceName }: Props) => {
+  const { data: alertManagerAlerts, isLoading: amAlertsIsLoading } =
```
```suggestion
  const { data: alertManagerAlerts = [], isLoading: amAlertsIsLoading } =
```
Might be easier than handling undefined later in the code :)

tomratcliffe committed:
Split handlers out into separate files

tomratcliffe committed:
Check for server received body

This is not ideal! Preference would be to have a more robust mock server
that responds to the received silence and appends it to a stateful list for the test
and then resets afterwards

konrad147 commented on public/app/features/alerting/unified/mocks/server/handlers.ts:
```diff
@@ -4,7 +4,60 @@
+
+///////////////////
+// Alertmanagers //
+///////////////////
```
Wouldn't be better to split these into multiple files?

tomratcliffe committed:
Rename silences RTKQ tag

tomratcliffe committed:
Add fallback for alertmanager alerts

konrad147 commented on public/app/features/alerting/unified/state/reducers.ts:
```diff
@@ -32,18 +29,13 @@ export const reducer = combineReducers({
   promRules: createAsyncMapSlice('promRules', fetchPromRulesAction, ({ rulesSourceName }) => rulesSourceName).reducer,
   rulerRules: createAsyncMapSlice('rulerRules', fetchRulerRulesAction, ({ rulesSourceName }) => rulesSourceName)
     .reducer,
-  silences: createAsyncMapSlice('silences', fetchSilencesAction, (alertManagerSourceName) => alertManagerSourceName)
```
🎉

tomratcliffe committed:
Check error state more reliably

konrad147 commented on public/app/features/alerting/unified/components/silences/SilencesEditor.tsx:
```diff
@@ -183,10 +189,14 @@ export const SilencesEditor = ({ silence, alertManagerSourceName }: Props) => {
   const userLogged = Boolean(config.bootData.user.isSignedIn && config.bootData.user.name);
 
+  if (getSilenceIsLoading) {
+    return null;
```
Would be nice to also cover error states, for example 404 for wrong ids

gillesdemey replied:
Agreed with Konrad on this point, do you want to tackle this in a follow-up PR or append to this one @tomratcliffe ?

tomratcliffe replied:
(I've added it in to show an alert when the silence can't be found, and have added a test for this as well. I would change the nav to add a button in this case but I would rather not let the changes spiral too much more in this PR if that's okay? WDYT?)

gillesdemey replied:
Agreed 👍

konrad147 commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -69,17 +96,39 @@ const SilencesTable = ({ silences, alertManagerAlerts, alertManagerSourceName }:
   }, [filteredSilencesExpired, alertManagerAlerts]);
 
+  if (isLoading || amAlertsIsLoading) {
+    return null;
```
Render `LoadingPlaceholder`

konrad147 commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -169,43 +218,34 @@ const useFilteredSilences = (silences: Silence[], expired = false) => {
-    display: flex;
-    flex-direction: row;
-    align-items: center;
+  addNewSilence: css({
```
This seems to be unused

tomratcliffe replied:
Thanks - that's the thing I don't like about the approach for styles, it doesn't get caught by linting 😁

konrad147 commented on public/app/features/alerting/unified/mocks/alertmanagerApi.ts:
```diff
@@ -38,3 +45,17 @@ export function mockAlertmanagerConfigResponse(
+export const alertmanagerAlertsListHandler = () =>
+  http.get('/api/alertmanager/:datasourceUid/api/v2/alerts', () =>
+    HttpResponse.json([
+      mockAlertmanagerAlert({
```
I think it would be nice if the defaults reflected real server responses.
I'd provide a few alerts of each type. If an alert is suppresed `silencedBy` should return an ID of a silence returned in by the silences handler

tomratcliffe replied:
I've linked them together a bit more now - using a common const to link the silence ID together. Is that okay with you for now?

tomratcliffe commented:
(https://github.com/grafana/grafana/issues/86981 created to capture existing bug, not fixing in this PR but will be addressed later on)

gillesdemey commented on public/app/features/alerting/unified/components/silences/SilencesEditor.tsx:
```diff
@@ -97,24 +95,24 @@ const getDefaultFormValues = (searchParams: URLSearchParams, silence?: Silence):
-export const SilencesEditor = ({ silence, alertManagerSourceName }: Props) => {
+export const SilencesEditor = ({ silenceId, alertManagerSourceName }: Props) => {
+  const history = useHistory();
+  const [getSilence, { data: silence, isLoading: getSilenceIsLoading }] =
```
I think it might be beneficial to add a comment for why we're going with a lazy query instead of a regular one (for it to work with blank forms)

gillesdemey commented on public/app/features/alerting/unified/Silences.test.tsx:
```diff
@@ -270,7 +239,9 @@ describe('Silence edit', () => {
-      await waitFor(() => expect(ui.editor.durationField.query()).not.toBeNull());
+      expect(await ui.editor.durationField.find()).toBeInTheDocument();
+
+      const postRequest = waitForServerRequest(silenceCreateHandler());
```
Should we not assert on `expect(saveButton).toBeDisabled()` and then `await waitFor(() => expect(saveButton).not.toBeDisabled())`? This is what I've settled on with the settings editor, but had to add a [`delay`](https://mswjs.io/docs/api/delay/) to the POST handler so the button switches disabled state properly.

tomratcliffe replied:
Here I've stuck with asserting that we were returned to the list view, which is what happens for a user on a successful silence creation. WDYT?

gillesdemey replied:
Oh that's also perfect, do we still need to call `waitForServerRequest` in that case?

konrad147 replied:
Would a simple array storing silences help in this case? Each `POST` could add a new item to the array.
Not sophisticated but should work

tomratcliffe replied:
We could do something along those lines, but it would have to be reset between each test using a jest hook, and I'm not sure how best to set it up in a way that scales right now, but also integrates cleanly with the current direction of our mocking approach.

An alternative I thought about was to make the `post` handler do a `server.use(..)` with an overridden list handler, but as we keep our handlers separated from the server (intentionally) it isn't super straightforward to implement right now.

The recommendation [here](https://github.com/mswjs/msw/discussions/964) is to use something like [mswjs/data](https://github.com/mswjs/data), or maintain state yourself (as you suggested).

Do you mind if I defer that decision to a separate PR/later work so we can review that implementation in more detail?

gillesdemey replied:
Right that sounds like a step too far for now – `waitForServerRequest` certainly has its use here so let's keep it for now :)

gillesdemey commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -29,12 +33,35 @@ export interface SilenceTableItem extends Silence {
+  const { data: alertManagerAlerts = [], isLoading: amAlertsIsLoading } =
+    alertmanagerApi.endpoints.getAlertmanagerAlerts.useQuery(
+      { amSourceName: alertManagerSourceName, filter: { silenced: true, active: true, inhibited: true } },
+      { pollingInterval: SILENCES_POLL_INTERVAL_MS }
```
nit: perhaps we should also add `refetchOnFocus` for our RTKQ hooks here? 20s is pretty slow and the API responses should be quite fast since the silences are kept in-memory in Alertmanager.

gillesdemey commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -69,17 +96,39 @@ const SilencesTable = ({ silences, alertManagerAlerts, alertManagerSourceName }:
+  }
+
+  if (isFetchError(error)) {
+    const errMessage = error?.message || 'Unknown error.';
```
prefer [`stringifyErrorLike`](https://github.com/grafana/grafana/blob/89d955fd9f8440eed9eb4bd2660ba1830c97f58d/public/app/features/alerting/unified/utils/misc.ts#L232-L239)

gillesdemey commented on public/app/features/alerting/unified/components/silences/SilencesTable.tsx:
```diff
@@ -69,17 +96,39 @@ const SilencesTable = ({ silences, alertManagerAlerts, alertManagerSourceName }:
           <SilencesFilter />
           <Authorize actions={[AlertmanagerAction.CreateSilence]}>
-            <div className={styles.topButtonContainer}>
+            <Stack justifyContent="end">
```
🙌

tomratcliffe commented on public/app/features/alerting/unified/api/alertSilencesApi.ts:
```diff
@@ -0,0 +1,65 @@
+      query: ({ datasourceUid, id }) => ({
+        url: `/api/alertmanager/${datasourceUid}/api/v2/silence/${id}`,
+      }),
+      providesTags: ['AlertmanagerSilences'],
```
I can add this in :+1: but for now we would still need to invalidate the "list" of silences as well so the UI refetches the table view properly. But it makes sense to start being more comprehensive about our cache tags here

tomratcliffe replied:
Done - let me know your thoughts

gillesdemey replied:
I've been going back and forth on this since revising the tags in the settings PR and I understand now that it was actually unnecessary for it to be separate tags but it doesn't hurt much either :D implementation LGTM!

tomratcliffe committed:
Add more fine grained tag invalidation/providing for Silences

tomratcliffe committed:
Address misc PR feedback

tomratcliffe committed:
Change test render method to use locationService

tomratcliffe committed:
Link silence test handlers together more clearly and test more cases

tomratcliffe commented on public/app/features/alerting/unified/api/alertSilencesApi.ts:
```diff
@@ -0,0 +1,65 @@
+        url: `/api/alertmanager/${datasourceUid}/api/v2/silence/${silenceId}`,
+        method: 'DELETE',
+      }),
+      invalidatesTags: ['AlertmanagerSilences'],
```
I've left this one as is, as we need to invalidate the entire tag of silences in this case anyway

tomratcliffe commented on public/test/test-utils.tsx:
Made some changes here to the test render method as we can't really get away without using `locationService`, but we do need to generate a fresh one for each test that uses it (otherwise you have to hack around it and remember to push a new location on each test, which is confusing)

tomratcliffe committed:
Tidy up error message handling

gillesdemey reviewed (approved):
code LGTM 🚀

konrad147 commented on public/app/features/alerting/unified/mocks/silences.ts:
```diff
@@ -0,0 +1,26 @@
+export const silenceGetHandler = () =>
+  http.get<{ uuid: string }>('/api/alertmanager/:datasourceUid/api/v2/silence/:uuid', ({ params }) => {
+    const { uuid } = params;
+    const matchingMockSilence = mockSilences.find((silence) => silence.id === uuid);
```
How to override the default silences?

tomratcliffe replied:
For now, all we need in order to be able to test is either fetching an existing one, or trying to get a non-existent one. 

So I haven't added in a method to configure and return a different one, but if we need to test another case, we just add it to the array of silences that we mock, make the ID of that silence a `const`, and use that in our tests as well. What do you think?

tomratcliffe committed:
Add test to check that a broken alertmanager will handle errors correctly

konrad147 reviewed (approved):
LGTM!

tomratcliffe committed:
Include alertmanager param on creation

tomratcliffe committed:
Add more robust error handling for stringifying API responses
